File: tests/zig/fib_loop.zig
const std = @import("std");
const profile = @import("profile.zig");

var result: i32 = 0;
var count: u32 = 6511134;

noinline fn fibI(n: u32) i32 {
    var last: i32 = 1;
    var cur: i32 = 0;
    var i: u32 = 0;
    while (i < n) : (i += 1) {
        const tmp = cur;
        cur +%= last;
        last = tmp;
    }
    return cur;
}

fn run(_: void) void {
    result = fibI(@as(*volatile u32, &count).*);
    std.mem.doNotOptimizeAway(result);
}

pub fn main(init: std.process.Init) !void {
    try profile.profile(init, "fibonacci loop", run, {});
    if (result != 1781508648) profile.fail("fibonacci loop: wrong result");
}

File: tests/cpp/fib_loop.cpp
#include "profile.h"

static volatile int count = 6511134;

PROFILE_NOINLINE int fibI(int n) {
    int last = 1;
    int cur = 0;
    for (int i = 0; i < n; ++i) {
        int tmp = cur;
        cur = (int) ((unsigned) cur + (unsigned) last);
        last = tmp;
    }
    return cur;
}

int main() {
    int result = 0;
    profile::run("fibonacci loop", [&] { result = fibI(count); profile::keep(result); });
    if (result != 1781508648) profile::fail("fibonacci loop: wrong result");
    return 0;
}
